### DISCUSSION TITLE
How to run a grpc and grpc-web server on different ports?

### DISCUSSION THREAD

**Discussion by itmecho** [hyperium/tonic#740] [Q&A] [answered]
I'm wondering how to run multiple tonic servers concurrently. I currently have this working:
```rust
use apis::user::v1::user_service_server::UserServiceServer;
use tonic::transport::Server;

mod srv {
    use apis::user::v1::{user_service_server::UserService, GetUserRequest, User};
    use tonic::{Request, Response};
    pub struct Srv {}

    #[tonic::async_trait]
    impl UserService for Srv {
        async fn get_user(
            &self,
            _request: Request<GetUserRequest>,
        ) -> Result<Response<User>, tonic::Status> {
            let mut u = User::default();
            u.name = "test".to_owned();
            Ok(Response::new(u))
        }
    }
}

async fn wait() {
    println!("server listening");
    tokio::signal::ctrl_c().await.ok();
    println!("shutdown complete");
}

#[tokio::main]
async fn main() -> Result<(), Box<dyn std::error::Error>> {
    let grpc_server = Server::builder()
        .add_service(UserServiceServer::new(srv::Srv {}))
        .serve_with_shutdown("0.0.0.0:50051".parse()?, wait());

    let grpc_web_server = Server::builder()
        .accept_http1(true)
        .add_service(tonic_web::enable(UserServiceServer::new(srv::Srv {})))
        .serve_with_shutdown("0.0.0.0:50052".parse()?, wait());

    let grpc_handle = tokio::spawn(grpc_server);
    let grpc_web_handle = tokio::spawn(grpc_web_server);

    let _ = tokio::try_join!(grpc_handle, grpc_web_handle)?;
    Ok(())
}
```

but I need to pass a db pool into the `Srv` object. Are there any examples of this pattern I can take a look at?

**Comment by davidpdrsn** [comment, marked as answer]
Wait I'm confused. You example already runs two servers on different ports...

To answer what appears to be your real question:

`Srv` is an ordinary struct so you can do

```rust
pub struct Srv { db_pool: DatabasePool }

// and then create it with
.add_service(UserServiceServer::new(srv::Srv { db_pool }))

// and use it like any other struct field
async fn get_user(
    &self,
    _request: Request<GetUserRequest>,
) -> Result<Response<User>, tonic::Status> {
    // access the pool
    self.db_pool.do_something();

    // ...
}
```

**Reply by itmecho** [reply]
Yea, sorry, I'm doing a bad job wording it =\

So that works for the first server but then the second fails because the db pool has already been moved into the first server. I was wondering if I can do something like this:
```rust
pub struct Srv { db_pool: DatabasePool }

// create srv
let srv = Srv { db_pool };

// Then add the services
grpc_web.add_service(UserServiceServer::new(&srv))
grpc_web_server.add_service(UserServiceServer::new(&srv))
```

or if I need to use the initial version but wrap the db_pool in an `Arc`?

**Reply by itmecho** [reply]
> Wait I'm confused. You example already runs two servers on different ports...
> 
> To answer what appears to be your real question:
> 
> `Srv` is an ordinary struct so you can do
> 
> ```rust
> pub struct Srv { db_pool: DatabasePool }
> 
> // and then create it with
> .add_service(UserServiceServer::new(srv::Srv { db_pool }))
> 
> // and use it like any other struct field
> async fn get_user(
>     &self,
>     _request: Request<GetUserRequest>,
> ) -> Result<Response<User>, tonic::Status> {
>     // access the pool
>     self.db_pool.do_something();
> 
>     // ...
> }
> ```

I've also just realised you're David Pedersen! Thank you for all the rust coding streams, they're really useful :D

**Reply by davidpdrsn** [reply]
Connection pools generally implement `Clone` in a way that is cheap, otherwise you can wrap it in an `Arc`. And yeah then you just clone the connection when passing it into the first service so it doesn't get moved.

**Reply by davidpdrsn** [reply]
> I've also just realised you're David Pedersen! Thank you for all the rust coding streams, they're really useful 😃

Happy to hear. I need to get back into streaming. Been spending too much time on open-source 😜

**Reply by itmecho** [reply]
> Connection pools generally implement Clone in a way that is cheap, otherwise you can wrap it in an Arc. And yeah then you just clone the connection when passing it into the first service so it doesn't get moved.

Thanks! That makes sense =]

> Happy to hear. I need to get back into streaming. Been spending too much time on open-source :stuck_out_tongue_winking_eye:

Either way you're giving back to the community :D

**Comment by Tudyx** [comment]
From what I understand, I suppose that it’s not possible to have only one gRPC server that will handle both gRPC web (HTTP1) and normal gRPC (HTTP2) at the same time?
